### PR TITLE
Add formatters

### PR DESCRIPTION
**Pull request by rusty-key**
While we are waiting for formatters implementation in `ocaml-lsp`, we can use port of formatters I used in my previous experiments

### CODE REVIEW

File: src/Esy.re [hunk 1]
```diff
@@ -127,7 +127,9 @@ let processDeps = (dependenciesJson, folder) => {
         | Some(bsPlatformVersionJson) =>
           switch (Js.Json.classify(bsPlatformVersionJson)) {
           | JSONString(bsPlatformVersion) =>
-            if (bsPlatformVersion->Semver.minVersion->Semver.satisfies(">=6.0.0")) {
+            if (bsPlatformVersion
+                ->Semver.minVersion
+                ->Semver.satisfies(">=6.0.0")) {
```
Comment (rusty-key) [inline src/Esy.re]: Sorry for unrelated changes, previously this file wasn't formatted properly.
Everything in this file below this point is just formatting

File: src/ProjectType.re [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+             };
+           } else {
+             Opam |> resolve;
+           };
+         } else {
+           reject(Failure("Not a valid esy/opam/bsb project"));
+         }
+       )
+  );
+};
```
Comment (rusty-key) [inline src/ProjectType.re]: Extraction from `LSP.re`

File: src/formatters/FormatterUtils.re [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+let getFormatterPath = formatter => {
+  let rootPath = Vscode.Workspace.rootPath;
+  ProjectType.detect(rootPath)
+  |> P.then_(projectType => {
+       let esy = Node.processPlatform == "win32" ? "esy.cmd" : "esy";
+       switch (projectType) {
+       | ProjectType.Esy(_) => P.resolve({j|$esy $formatter|j})
+       | Bsb(_) =>
+         P.resolve({j|cd $rootPath/.vscode/esy && $esy $formatter|j})
+       | Opam => P.resolve({j|opam exec $formatter|j})
```
Comment (rusty-key) [inline src/formatters/FormatterUtils.re]: This is probably very silly, I am open to suggestions

File: src/formatters/Refmt.re [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    {
+      "provideDocumentFormattingEdits": () => {
+        switch (Vscode.Window.activeTextEditor) {
+        | None => Js.Promise.resolve([||])
+        | Some(textEditor) =>
+          let id = Uuid.v4();
+          let tempFileName =
+            Node.Path.join([|
+              Node.Os.tmpdir(),
+              {j|vscode-reasonml-refmt-$id.re|j},
```
Comment (prometheansacrifice) [inline src/formatters/Refmt.re]: The extension must be borrowed from the input file. For rei files, the formatting would fail otherwise

Comment (rusty-key) [reply src/formatters/Refmt.re]: I am trying to remember now why did I go with temp file at all

**Comment by prometheansacrifice** [comment]
Bummer. `ocamlformat` failed to build on Windows

**Comment by rusty-key** [comment]
@prometheansacrifice, symlinks issue. Not sure how it can be solved properly. I heard that it could be related to commands not be executed by administrator, but I don't know if it is the case

**Comment by rusty-key** [comment]
For example, it is solved in Angstrom by removing symlinks: 
https://github.com/inhabitedtype/angstrom/pull/153

**Comment by prometheansacrifice** [comment]
Great find. Thanks! We might have to resort to doing the same

**Comment by wokalski** [comment]
Maybe this should be merged without windows support and track windows in a separate issue?

**Comment by prometheansacrifice** [comment]
The CI ran fine. The release job needs to run only on master. Merging this.
